### ISSUE TITLE
Connect audioStreamer to ASR agent

### ISSUE THREAD

**Issue by eduongAZ** [ml4ai/tomcat#289]
Test session: April 12, 2020

Before starting the trial, speechAgent connected to 2 client maps. During the trial, the speechAgent still only connected to 2 client maps, and did not connect to the 3rd client map (on Tiger).

We need a consistent procedure for Clientmaps to connect to speechAgent.

**Comment by vincentraymond-ua** [comment]
I think this issue should probably be renamed to "ClientMap does not connect to ASRAgent", since the issue would be on the ClientMap side.

**Comment by vincentraymond-ua** [comment]
Also, the ClientMaps should not be able to connect to the ASRAgent before the Trial starts. This could affect the data pipeline, as there are some components that assume the audio starts at trial start. The Testbed used to prevent this from happening.

**Comment by vincentraymond-ua** [comment]
I will try testing locally to see if I can recreate the issue.

**Comment by eduongAZ** [comment]
@vincentraymond-ua What would be your recommended steps for setting up the testbed, connecting to clientmap, launching the agent, and start a trial such that client maps connect to speechAnalyzer agent?

**Comment by vincentraymond-ua** [comment]
There shouldn't really be many additional steps to use the ASRAgent and SpeechAnalyzer. Here are the steps that I usually do:

1. Bring down the testbed with testbed_down.sh
2. Bring up the testbed with testbed_up_core.sh -a -w
3. Open MalmoControl in a web browser
4. Open ClientMaps in web browser and login
5. Start trial as normal 
6. Allow microphone access in ClientMap windows

**Comment by adarshp** [comment]
Update: This will be tested tomorrow. If it works, someone (maybe @vincentraymond-ua ?) should add it to the ml4ai/tomcat `docs` folder.

**Comment by eduongAZ** [comment]
The clientmap connects to the ASRAgent following these steps:

1. Launch testbed with `sudo ./testbed_up_code.sh -a -w`
2. Open MalmoControl in web browser.
3. Open ClientMaps in web browser and login
4. Create and start trial.
5. Toggle the speaker icon and allow microphone access in ClientMap windows.

It still seems uncertain whether the ClientMap will connect to the testbed. We will follow the steps outlined above for a few times to ensure it is reliable.

**Comment by eduongAZ** [comment]
Clientmap failed to connect to ASR agent despite following the above steps. During test session on April 26, 2022, only one client map connected to the ASR agent when following the same procedure for all 3 imacs. However, after connecting the ASR agent, the agent immediately closed Google session, disconnecting the Clientmap.

**Comment by adarshp** [comment]
Given the issues we are facing with the ClientMap audio capture, I wonder if we should simply bypass the browser for the audio capture - this is feasible for us since we are not trying to do remote experiments and we have full control over the client machines.

Specifically, how about this plan: we do not use the ClientMap's audio capture functionality. Instead, we use our own [audioStreamer](https://github.com/ml4ai/tomcat/tree/master/exe/audioStreamer) program for streaming audio to the server for processing.

I think we will need to the do the following for this to work:

1. Modify the audioStreamer program to interface with the message bus (I recommend using the Paho C++ client library) so that we can start/stop streaming using messages (could use mission start/stop messages)
2. Modify the ASR_Agent program to have the ability to record .wav files for individual participants (what the AsistDataIngester is currently doing). We already have a way to record raw audio, we could use`libsndfile` to make .wav files happen.
3. We need to test the modifications :)

**Comment by vincentraymond-ua** [comment]
I would support this option, the audioStreamer program is pretty lightweight and easy to use, and the modifications we would need to make would be relatively simple.

My main concerns would be:

1.  Are we able to disable audio streaming through the ClientMap entirely? If it works sporadically, then we could see a situation where we have three audio streams connected to the ASR_Agent through the audioStreamer, and 1-3 streams connected through the ClientMap. This could cause issues with the data generated through the ASR_Agent and downstream components.
2.  The fact that we are having these audio streaming issues and other teams are not could imply that we have a larger issue with our Testbed install that could cause further errors down the line.

**Comment by vincentraymond-ua** [comment]
I think we can work around the first point by modifying the testbed/Local/AsistDataIngester/appsettings.json file.

**Comment by vincentraymond-ua** [comment]
Would it make sense to have the audioStreamer record the .wav files instead of the ASR_Agent? That would simulate the AsistDataIngester more accurately, since the audio would be preserved if the ASR_Agent crashes. The audio streamer has access to the same raw audio chunks that the ASR_Agent does, so we would only need `libsndfile` convert those to .wav format.

**Comment by adarshp** [comment]
@vincentraymond-ua Sure that will work for the Lang Lab setup since we are using NFS. However, I was hoping to use the audioStreamer for the SensorLab projects too, and we don't want to record the .wav files onto the Raspberry Pis themselves, but rather on the central audio ingestion and processing server.

**Comment by adarshp** [comment]
But if it is more expedient to do the NFS route you can go ahead and implement the recording by audioStreamer and have a command line option or something to disable the recording.

**Comment by kobus-barnard** [comment]
A command line option might be best if we want both ways. NFS simplifies things, and it seems to be working as it should.

**Comment by vincentraymond-ua** [comment]
Agreed, I think we can go with the command line option in the audioStreamer to enable/disable recording for now. We can later add the same recording functionality to the speechAnalyzer for the SensorLab project.

**Comment by eduongAZ** [comment]
During the dry run, ClientMaps connected to ASR agent when using the following procedure:

1. Launch Minecraft on ToM and iMacs.
2. Launch testbed with option -i
3. Log in on ClientMaps.
4. Toggle on for all speaker icons on the ClientMaps.
5. Launch the mission on Malmo Control.
6. Toggle off and on for all speaker icons on the ClientMaps.
7. After this step, the ASR speech agent acknowledges the connections of the ClientMaps.
8. Start the mission.

**Comment by vincentraymond-ua** [comment]
Here is the report from our lab tests today.

Success:
- Disable ClientMap audio streaming
- Build and launch audioStreamer
- Connect audioStreamer to MQTT bus on Tom

Failures:
- Could not connect to ASR_Agent container running websocket server
- Could not get port_forwarder script working on iMacs

On Monday, we will test this further. The audioStreamer itself appears to be functioning, we just have to facilitate the connection between it and the ASR_Agent.

**Comment by vincentraymond-ua** [comment]
We were able to get the audioStreamer to connect to the ASR_Agent without the need for port-forwarding. I believe we must have had a typo in the IP address when we were testing previously.

All that is left to do for this task is to have the audioStreamer incorporated into the Uber script.

**Comment by paulosoaresua** [comment]
@vincentraymond-ua, what's the incantation to start the audio streamer and in which phase of the experiment must it be fired?

**Comment by vincentraymond-ua** [comment]
The audioStreamer should be launched after the testbed is brought up. It will listen to the MQTT message bus for a trial start message, and then connect to the ASR_Agent and start streaming/recording audio. 

The incantation is: 
`./audioStreamer --player_name $PLAYER_NAME --ws_host $TOM_IP  --use_mqtt true --mqtt_host $TOM_IP`

$PLAYER_NAME - The name of the player from the Malmo Minecraft Client
$TOM_IP - The IP address of the Tom machine

**Comment by eduongAZ** [comment]
The audioStreamer will be tested in the dry run on June 7.

**Comment by vincentraymond-ua** [comment]
@eduongAZ - Caleb and I tested the audioStreamer on 5/17 and were able to get it working with the invocation in my comment above to Paulo. I think what we are testing at the dry run today is the sensitivity of the microphones with the ASRAgent in reference to this issue: #319

**Comment by eduongAZ** [comment]
Thank you for the update. The issue is not closed so I was confused on whether the problem was resolved and tested in the lab. Since it has both been resolved and tested in the lab, I am closing this issue.

**Comment by eduongAZ** [comment]
@vincentraymond-ua We tested the audioStreamer, and it seems that, when looking into "agent/final/asr" topic, there are multiple duplicates of the same speech being posted. Is this because ClientMap audio streaming is also working? If so, how can we disable audio streaming feature of the ClientMap (besides clicking the speaker icon on the ClientMap, because the agent automatically turns the speaker icon on)?

**Comment by vincentraymond-ua** [comment]
Hi Eric, 

Could this be related to #319 ? We were seeing duplicate messages previously because the microphone was very sensitive.

To disable the ClientMap audio streaming, I believe we determined that you need to modify testbed/Local/AsistDataIngester/appsettings.json.

Toggle line 22 ("DisableSpeechToText") from "true" to "false".

**Comment by adarshp** [comment]
These were not duplicate messages - just the same text occurring multiple times in the same message in the `.data.text` field.

**Comment by vincentraymond-ua** [comment]
That's strange, I am not seeing this behavior on my end when using the audioStreamer. I don't think Caleb and I saw this behavior either when initially testing the audioStreamer at the lab. 

The data.text field comes directly from Google, so the audio that is getting sent out must actually have these duplicated utterances. 

Are there any programs running that could cause an echo or feedback in the mics? Maybe Mumble?

@eduongAZ - When you are in the lab again, could you provide an example of a message with this issue?

**Comment by eduongAZ** [comment]
It looks like the same text occurring multiple times in the same message is caused by very sensitive microphones.

![Image](https://user-images.githubusercontent.com/86935179/174897426-fe904a88-03dc-4db6-88bd-2b757ea48520.png)
![Image](https://user-images.githubusercontent.com/86935179/174897425-c3088d64-df3a-4f38-87c8-d447ba86d45a.png)
![Image](https://user-images.githubusercontent.com/86935179/174897740-0c97f677-95a1-415b-9117-2b54752350d0.png)
